src/mito2/src/worker/handle_truncate.rs

```diff
@@ -28,24 +29,65 @@ impl<S: LogStore> RegionWorkerLoop<S> {
     pub(crate) async fn handle_truncate_request(
         &mut self,
         region_id: RegionId,
+        req: RegionTruncateRequest,
         mut sender: OptionOutputTx,
     ) {
         let Some(region) = self.regions.writable_region_or(region_id, &mut sender) else {
             return;
         };
+        let version_data = region.version_control.current();
 
-        info!("Try to truncate region {}", region_id);
+        let truncate = match req {
+            RegionTruncateRequest::All => {
+                info!("Try to fully truncate region {}", region_id);
 
-        let version_data = region.version_control.current();
-        let truncated_entry_id = version_data.last_entry_id;
-        let truncated_sequence = version_data.committed_sequence;
+                let truncated_entry_id = version_data.last_entry_id;
+                let truncated_sequence = version_data.committed_sequence;
 
-        // Write region truncated to manifest.
-        let truncate = RegionTruncate {
-            region_id,
-            truncated_entry_id,
-            truncated_sequence,
+                // Write region truncated to manifest.
+                RegionTruncate {
+                    region_id,
+                    kind: TruncateKind::All {
+                        truncated_entry_id,
+                        truncated_sequence,
+                    },
+                }
+            }
+            RegionTruncateRequest::ByTimeRanges { time_ranges } => {
+                debug!(
+                    "Try to partially truncate region {} by time ranges: {:?}",
+                    region_id, time_ranges
+                );
+                // find all files that are fully contained in the time ranges
+                let mut files_to_truncate = Vec::new();
+                for level in version_data.version.ssts.levels() {
+                    for file in level.files() {
+                        let file_time_range = file.time_range();
+                        // TODO(discord9): This is a naive way to check if is contained, we should
+                        // optimize it later.
+                        let is_subset = time_ranges.iter().any(|(start, end)| {
+                            file_time_range.0 >= *start && file_time_range.1 <= *end
+                        });
```

Comment:
I think you may want to merge contiguous time ranges first, otherwise some files are not selected. For example, if the input time ranges are [(1, 3), (3, 5)], then the file with time range (2, 4) will not be considered as "subset" here, while it actually should.

Reply:
> I think you may want to merge contiguous time ranges first, otherwise some files are not selected. For example, if the input time ranges are [(1, 3), (3, 5)], then the file with time range (2, 4) will not be considered as "subset" here, while it actually should.

don't want to over-complex stuff here, this `truncate foo file range` thing will be only basically used by flow, so complex interval merging stuff should happen there not here for better debug&simplicity

Reply:
ok, then this should be added to the comments